### PR TITLE
New processor extract_array

### PR DESCRIPTION
**Pull request by adriansr**
This PR adds a new processor, `extract_array`, that allows to map the contents of an array field into other fields:

```yaml
 - extract_array:
     field: array_field
     mappings:
         event.type: 0
         source.ip: 1
         destination.ip: 2
         network.transport: 3
```

I needs this to use with the `decode_csv_processor` (#11753) as in:
```yaml
processors:
 - decode_csv_fields:
     fields:
     message: csv
 - extract_array:
     field: csv
     mappings:
         observer.serial_number: 2
         palo_alto.message_type: 3
 - extract_array:
     field: csv
     when:
         equals:
             palo_alto.message_type: TRAFFIC
     mappings:
         source.ip: 7
         destination.ip: 8
         source.user.name: 12
         destination.user.name: 13
         network.application: 14
         source.port: 24
         destination.port: 25
         network.transport: 29
         event.action: 30
         network.bytes: 31
         source.bytes: 32
         destination.bytes: 33
         network.packets: 34
         event.duration: 36
         source.geo.country_iso_code: 41
         destination.geo.country_iso_code: 42
         source.packets: 44
         destination.packets: 45
```

### CODE REVIEW

File: libbeat/processors/extract_array/extract_array.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+				continue
+			}
+			return &saved, errors.Wrapf(err, "failed setting field %s", mapping.to)
+		}
+	}
+	return event, nil
+}
+
+func (f *extractArrayProcessor) String() (r string) {
+	return fmt.Sprintf("extract_array={field:%s, mappings:%v}", f.config.Field, f.mappings)
```
Comment (andrewkroh) [inline libbeat/processors/extract_array/extract_array.go]: IIRC the other processor use name=[field=x, other=y] as their format. I might not be recalling this correctly.

File: libbeat/processors/extract_array/extract_array.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func clone(value interface{}) interface{} {
+	// TODO: This is dangerous but done by most processors.
+	//       Otherwise need to reflect value and deep copy lists / map types.
+	switch v := value.(type) {
+	case common.MapStr:
```
Comment (andrewkroh) [inline libbeat/processors/extract_array/extract_array.go]: Is is possible after going through the event normalization for one of the values to be an `map[string]interface{}`? Perhaps just to be defensive you should add a case statement for it and cast it to `common.MapStr` and then do the shallow `Clone()`.

**Comment by adriansr** [comment]
jenkins, test this

File: libbeat/processors/extract_array/extract_array.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+		return v.Clone()
+	case map[string]interface{}:
+		return common.MapStr(v).Clone()
+	}
+	return value
+}
```
Comment (urso) [inline libbeat/processors/extract_array/extract_array.go]: I presume we use clone, so modifications don't affect the original data within the array? In this case also look out for arrays. Some processors like `add_tags` can modify arrays.

File: libbeat/docs/processors-using.asciidoc [hunk 4]
```diff
@@ -206,10 +206,13 @@ The supported processors are:
...
  * <<community-id,`community_id`>>
 ifeval::[("{beatname_lc}"=="filebeat") or ("{beatname_lc}"=="journalbeat")]
-* <<decode-csv-fields,`decode_csv_fields`>>
+ * <<decode-csv-fields,`decode_csv_fields`>>
 endif::[]
  * <<decode-json-fields,`decode_json_fields`>>
  * <<dissect, `dissect`>>
+ifeval::[("{beatname_lc}"=="filebeat") or ("{beatname_lc}"=="journalbeat")]
+ * <<extract-array,`extract_array`>>
+endif::[]
```
[the remark this answers is not in the document]
Comment (adriansr) [reply libbeat/docs/processors-using.asciidoc]: We discussed this during the Beats contributors meeting and I recall that was the agreement. Maybe that only makes sense for the CSV processor? I will make it available to all, makes more sense.

[the remark this answers is not in the document]
Comment (dedemorton) [reply libbeat/docs/processors-using.asciidoc]: sounds like my comment is no longer relevant but wanted to mention that complex conditionals (with `and` and `or`) aren't supported by asciidoctor. To get around this, I'm using ifdef and ifndef statements that look for attributes defined in the book. If you want to see what that looks like, see the changes in https://github.com/elastic/beats/pull/12006.

File: libbeat/docs/processors-using.asciidoc [hunk 5]
```diff
@@ -1695,3 +1698,48 @@ thrown.
...
+     mappings:
+        source.ip: 0
+        destination.ip: 1
+        network.transport: 2
+-----------------------------------------------------
+
+The following settings are supported:
+
+`field`:: The array field whose elements are to be extracted.
+`mappings`:: A field name to array index mapping. Use 0 for the first element in
```
Comment (dedemorton) [inline libbeat/docs/processors-using.asciidoc]: This sentence is a bit hard to parse on first and second read.  How about: 
```suggestion
`mappings`:: Maps each field name to an array index. Use 0 for the first element in
```

File: libbeat/docs/processors-using.asciidoc [hunk 6]
```diff
@@ -1695,3 +1698,48 @@ thrown.
...
+             the array. Multiple fields can be mapped to the same array element.
+`ignore_missing`:: (Optional) Whether to ignore events where the array field is
+                   missing. The default is false, which will fail processing of
+                   an event if the specified field does not exist. Set it to
+                   `true` to ignore this condition.
+`overwrite_keys`:: Whether the target fields specified in the mapping are
+                   overwritten if they already exist. The default is `false`,
+                   which will fail processing if a target field already exists.
+`fail_on_error`:: (Optional) If set to `true`, in case of an error, changes to
```
Comment (dedemorton) [inline libbeat/docs/processors-using.asciidoc]:
```suggestion
`fail_on_error`:: (Optional) If set to `true` and an error happens, changes to
```

File: libbeat/docs/processors-using.asciidoc [hunk 7]
```diff
@@ -1695,3 +1698,48 @@ thrown.
...
+                  the event are reverted and the original event is returned. If
+                  set to false, processing continues if an error happens
```
Comment (dedemorton) [inline libbeat/docs/processors-using.asciidoc]:
```suggestion
                  set to `false`, processing continues despite errors.
```

File: libbeat/docs/processors-using.asciidoc [hunk 8]
```diff
@@ -1695,3 +1698,48 @@ thrown.
...
+             the array. Multiple fields can be mapped to the same array element.
+`ignore_missing`:: (Optional) Whether to ignore events where the array field is
+                   missing. The default is false, which will fail processing of
+                   an event if the specified field does not exist. Set it to
+                   `true` to ignore this condition.
+`overwrite_keys`:: Whether the target fields specified in the mapping are
+                   overwritten if they already exist. The default is `false`,
+                   which will fail processing if a target field already exists.
+`fail_on_error`:: (Optional) If set to `true`, in case of an error, changes to
+                  the event are reverted and the original event is returned. If
```
Comment (dedemorton) [inline libbeat/docs/processors-using.asciidoc]:
```suggestion
                  the event are reverted, and the original event is returned. If
```

File: libbeat/docs/processors-using.asciidoc [hunk 9]
```diff
@@ -1695,3 +1698,48 @@ thrown.
...
+                  set to false, processing continues if an error happens
+                  during processing. Default is `true`.
```
Comment (dedemorton) [inline libbeat/docs/processors-using.asciidoc]:
```suggestion
                  Default is `true`.
```

File: libbeat/processors/extract_array/extract_array.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+			if !f.config.FailOnError {
+				continue
+			}
+			return &saved, errors.Errorf("index %d exceeds length of %d when processing mapping for field %s", mapping.from, n, mapping.to)
+		}
+		cell := array.Index(mapping.from)
+		// checking for CanInterface() here is done to prevent .Interface() from
+		// panicking, but it can only happen when value points to a private
+		// field inside a struct.
+		if !cell.CanInterface() || (f.config.OmitEmpty && isEmpty(cell)) {
```
[the remark this answers is not in the document]
Comment (adriansr) [reply libbeat/processors/extract_array/extract_array.go]: Added the IsValid() check, thanks! I don't see an IsZero() though.

[the remark this answers is not in the document]
Comment (urso) [reply libbeat/processors/extract_array/extract_array.go]: Sorry, I did mean IsNil

[the remark this answers is not in the document]
Comment (adriansr) [reply libbeat/processors/extract_array/extract_array.go]: IsNil() is used inside isEmpty for some types (as otherwise it can panic). I added one test case with nil values to make sure there's no problem.

**Comment by adriansr** [comment]
jenkins, test this

**Comment by adriansr** [comment]
Rebased to fix CI
